File: non-streaming-request-over-ten-minutes/python/anthropic_wall_clock_preflight.py
"""Estimate whether a non-streaming Claude call can finish inside 10 minutes.

Read only, with one deliberate exception. Nothing here creates a completion:
where a call path names a payload file, that body goes to
/v1/messages/count_tokens, which is free, creates no object, generates no
output and is not billed. It is used to turn the input into prefill seconds.
Everything else is a GET, and /v1/messages is never called.

The repair is a transport change and it is printed. Moving a production call
path onto streaming changes error handling and back pressure, which is a
decision, not an audit's side effect.
"""
import argparse
import json
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("anthropic_wall_clock_preflight")

API = "https://api.anthropic.com/v1"
VERSION = "2023-06-01"

# The documented ceiling for a single non-streaming Messages request.
WALL_CLOCK = 600.0

# Starting figures, both meant to be replaced with your own measurements.
# Generation is the one that decides the answer; prefill is fast enough that it
# only matters on very large inputs, and proving that is half the point.
DEFAULT_TPS = 55.0
DEFAULT_PREFILL_TPS = 6000.0

# Client timeouts are not expressed in the same unit across SDKs, and a number
# copied from one language's example into another's constructor is the quiet
# half of this note.
SDK_TIMEOUT_UNITS = {
    "python": ("seconds", 1.0),
    "ruby": ("seconds", 1.0),
    "php": ("seconds", 1.0),
    "typescript": ("milliseconds", 0.001),
    "javascript": ("milliseconds", 0.001),
    "node": ("milliseconds", 0.001),
    "go": ("a time.Duration", 1.0),
    "java": ("a Duration", 1.0),
    "csharp": ("a TimeSpan", 1.0),
}
MILLISECOND_SDKS = ("typescript", "javascript", "node")

SAMPLING_ONLY = ("max_tokens", "stream", "temperature", "top_p", "top_k",
                 "stop_sequences", "metadata", "service_tier")

FINDINGS = ("over-wall-clock-not-streaming", "over-client-timeout",
            "near-wall-clock-not-streaming")


def duration(seconds):
    """Seconds as minutes and seconds. Pure."""
    total = int(max(0.0, float(seconds or 0)))
    return "%dm %02ds" % (total // 60, total % 60)


def generation_seconds(max_tokens, tps=DEFAULT_TPS):
    """How long it takes to write max_tokens output tokens. Pure.

    This is the number the whole note turns on, and it has nothing to do with
    the size of the prompt. Thinking tokens are output tokens, so an effort
    setting moves it too.
    """
    rate = float(tps or 0)
    if rate <= 0:
        return 0.0
    return max(0, int(max_tokens or 0)) / rate


def prefill_seconds(input_tokens, prefill_tps=DEFAULT_PREFILL_TPS):
    """How long it takes to read the input. Pure.

    Kept separate and reported separately because it is almost always small,
    and the point of measuring it is to stop people shortening prompts to fix a
    problem that lives entirely on the output side.
    """
    rate = float(prefill_tps or 0)
    if rate <= 0:
        return 0.0
    return max(0, int(input_tokens or 0)) / rate


def timeout_seconds(sdk, value):
    """A client timeout in seconds, whatever unit the SDK takes. Pure.

    None when the SDK is unknown, because guessing the unit is precisely the
    mistake this function exists to catch.
    """
    if value is None:
        return None
    unit = SDK_TIMEOUT_UNITS.get(str(sdk or "").strip().lower())
    if unit is None:
        return None
    try:
        return float(value) * unit[1]
    except (TypeError, ValueError):
        return None


def unit_suspicion(sdk, value):
    """True when a timeout looks written in the wrong unit. Pure.

    600 in the TypeScript client is six hundred milliseconds, not ten minutes.
    Nobody chooses a sub-second timeout for an LLM call on purpose, so anything
    under a second on a millisecond SDK is a number copied from a seconds-based
    example.
    """
    seconds = timeout_seconds(sdk, value)
    if seconds is None:
        return False
    return str(sdk or "").strip().lower() in MILLISECOND_SDKS and seconds < 1.0


def safe_max_tokens(tps=DEFAULT_TPS, wall_clock=WALL_CLOCK, prefill=0.0):
    """The largest max_tokens that still finishes inside the ceiling. Pure.

    The number to put in the config, as opposed to the model's cap, which is
    the number that fits in the request.
    """
    rate = float(tps or 0)
    room = max(0.0, float(wall_clock or 0) - max(0.0, float(prefill or 0)))
    if rate <= 0:
        return 0
    return int(room * rate)


def verdict(seconds, streams, timeout_s=None, wall_clock=WALL_CLOCK, near=0.8):
    """Classify one call path against the clock. Pure. (state, detail).

    Order matters. The wall clock is checked before the client timeout, because
    a non-streaming request past ten minutes fails on the far side whatever the
    client is configured to wait for, and raising the client timeout is both
    the first repair people reach for and the one that does nothing.
    """
    shape = "%s of generation estimated" % duration(seconds)

    if not streams and seconds > wall_clock:
        return ("over-wall-clock-not-streaming",
                "%s on a non-streaming path, past the %s ceiling. That is a 504 "
                "timeout_error, or no response at all when an intermediate hop "
                "drops the idle connection first. Raising the client timeout "
                "does not move it." % (shape, duration(wall_clock)))
    if timeout_s is not None and seconds > timeout_s:
        return ("over-client-timeout",
                "%s against a client timeout of %s, so the client gives up "
                "before the API is finished." % (shape, duration(timeout_s)))
    if not streams and seconds >= wall_clock * near:
        return ("near-wall-clock-not-streaming",
                "%s on a non-streaming path, inside %.0f%% of the %s ceiling. "
                "One unusually long answer crosses it."
                % (shape, near * 100, duration(wall_clock)))
    if streams and seconds > wall_clock:
        return ("streams-past-ten-minutes",
                "%s, and the path streams, so the connection never goes idle "
                "and the ceiling does not apply. Worth the Message Batches API "
                "if nobody is waiting on it." % shape)
    return ("within-budget", "%s." % shape)


def get(session, path):
    r = session.get(API + path, timeout=30)
    if r.status_code in (401, 403):
        raise SystemExit("%d from Anthropic: ANTHROPIC_API_KEY has to be a "
                         "workspace key" % r.status_code)
    r.raise_for_status()
    return r.json()


def count_input(session, payload_path):
    """The one non-GET call, and it neither creates nor bills anything.

    The counting endpoint returns an input_tokens number for free. Here that
    number is converted straight into seconds of prefill; it is not compared
    against any ceiling, which is a different note.
    """
    with open(payload_path, "r", encoding="utf-8") as fh:
        body = json.load(fh)
    trimmed = {k: v for k, v in body.items() if k not in SAMPLING_ONLY}
    r = session.post(API + "/messages/count_tokens", json=trimmed, timeout=60)
    r.raise_for_status()
    return int((r.json() or {}).get("input_tokens") or 0)


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--config", required=True,
                    help="JSON file of call paths: "
                         '{"name": {"model": ..., "max_tokens": ..., '
                         '"stream": false, "sdk": "typescript", '
                         '"timeout": 600, "payload": "body.json"}}')
    ap.add_argument("--tps", type=float, default=DEFAULT_TPS,
                    help="observed output tokens per second (default 55)")
    ap.add_argument("--prefill-tps", type=float, default=DEFAULT_PREFILL_TPS,
                    help="observed input tokens per second (default 6000)")
    ap.add_argument("--show-all", action="store_true",
                    help="also print paths comfortably inside the clock")
    args = ap.parse_args()

    key = os.environ.get("ANTHROPIC_API_KEY")
    if not key:
        log.error("set ANTHROPIC_API_KEY to a workspace key")
        return 2

    with open(args.config, "r", encoding="utf-8") as fh:
        paths = json.load(fh)

    session = requests.Session()
    session.headers.update({"x-api-key": key, "anthropic-version": VERSION,
                            "content-type": "application/json"})

    caps = {}
    bad = 0

    for name in sorted(paths):
        entry = paths[name] or {}
        model_id = str(entry.get("model") or "")
        streams = bool(entry.get("stream"))
        sdk = entry.get("sdk")

        if model_id and model_id not in caps:
            caps[model_id] = get(session, "/models/" + model_id).get("max_tokens")

        input_tokens = int(entry.get("input_tokens") or 0)
        if entry.get("payload"):
            input_tokens = count_input(session, entry["payload"])

        prefill = prefill_seconds(input_tokens, args.prefill_tps)
        seconds = prefill + generation_seconds(entry.get("max_tokens"), args.tps)
        client = timeout_seconds(sdk, entry.get("timeout"))

        state, detail = verdict(seconds, streams, client)
        line = "%-30s %-16s %s" % (state, name, detail)
        if state in FINDINGS:
            bad += 1
            log.warning(line)
        elif state == "streams-past-ten-minutes":
            log.info(line)
        elif args.show_all:
            log.info(line)

        if unit_suspicion(sdk, entry.get("timeout")):
            bad += 1
            log.warning("%-30s %-16s timeout %s on the %s client is %.1fs, not "
                        "%s: that unit is milliseconds",
                        "timeout-unit-mistake", name, entry.get("timeout"), sdk,
                        client or 0.0, duration(entry.get("timeout") or 0))

        if state in ("over-wall-clock-not-streaming",
                     "near-wall-clock-not-streaming"):
            log.warning("  at %.0f tok/s the largest max_tokens that finishes "
                        "inside the ceiling is %d",
                        args.tps, safe_max_tokens(args.tps, WALL_CLOCK, prefill))
            cap = caps.get(model_id)
            if cap:
                log.warning("  this model allows %d output tokens, which is %s "
                            "on one call", cap,
                            duration(generation_seconds(cap, args.tps)))
            log.warning("  repair: stream it. .stream() plus "
                        ".get_final_message() returns the identical Message "
                        "object with no event handling, and the connection "
                        "never goes idle. For latency tolerant work use the "
                        "Message Batches API, which has no such clock. Printed, "
                        "not applied.")

    log.info("%d path(s) checked, %d finding(s)", len(paths), bad)
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())

File: non-streaming-request-over-ten-minutes/node/anthropic-wall-clock-preflight.mjs
/**
 * Estimate whether a non-streaming Claude call can finish inside 10 minutes.
 *
 * Read only, with one deliberate exception. Nothing here creates a completion:
 * where a call path names a payload file, that body goes to
 * /v1/messages/count_tokens, which is free, creates no object, generates no
 * output and is not billed. It is used to turn the input into prefill seconds.
 * Everything else is a GET, and /v1/messages is never called.
 *
 * The repair is a transport change and it is printed.
 */
import { readFile } from 'node:fs/promises';

const API = 'https://api.anthropic.com/v1';
const VERSION = '2023-06-01';

const WALL_CLOCK = 600;
const DEFAULT_TPS = 55;
const DEFAULT_PREFILL_TPS = 6000;

const SDK_TIMEOUT_UNITS = {
  python: ['seconds', 1],
  ruby: ['seconds', 1],
  php: ['seconds', 1],
  typescript: ['milliseconds', 0.001],
  javascript: ['milliseconds', 0.001],
  node: ['milliseconds', 0.001],
  go: ['a time.Duration', 1],
  java: ['a Duration', 1],
  csharp: ['a TimeSpan', 1],
};
const MILLISECOND_SDKS = new Set(['typescript', 'javascript', 'node']);

const SAMPLING_ONLY = new Set(['max_tokens', 'stream', 'temperature', 'top_p',
  'top_k', 'stop_sequences', 'metadata', 'service_tier']);

const FINDINGS = new Set(['over-wall-clock-not-streaming', 'over-client-timeout',
  'near-wall-clock-not-streaming']);

/** Seconds as minutes and seconds. Pure. */
export function duration(seconds) {
  const total = Math.trunc(Math.max(0, Number(seconds || 0)));
  return `${Math.floor(total / 60)}m ${String(total % 60).padStart(2, '0')}s`;
}

/** How long it takes to write maxTokens output tokens. Pure. */
export function generationSeconds(maxTokens, tps = DEFAULT_TPS) {
  const rate = Number(tps || 0);
  if (rate <= 0) return 0;
  return Math.max(0, Math.trunc(maxTokens || 0)) / rate;
}

/** How long it takes to read the input. Pure. Reported separately on purpose. */
export function prefillSeconds(inputTokens, prefillTps = DEFAULT_PREFILL_TPS) {
  const rate = Number(prefillTps || 0);
  if (rate <= 0) return 0;
  return Math.max(0, Math.trunc(inputTokens || 0)) / rate;
}

/**
 * A client timeout in seconds, whatever unit the SDK takes. Pure.
 * Null when the SDK is unknown, because guessing the unit is the mistake this
 * function exists to catch.
 */
export function timeoutSeconds(sdk, value) {
  if (value === null || value === undefined) return null;
  const unit = SDK_TIMEOUT_UNITS[String(sdk ?? '').trim().toLowerCase()];
  if (!unit) return null;
  const n = Number(value);
  return Number.isFinite(n) ? n * unit[1] : null;
}

/**
 * True when a timeout looks written in the wrong unit. Pure.
 * 600 in the TypeScript client is six hundred milliseconds, not ten minutes.
 */
export function unitSuspicion(sdk, value) {
  const seconds = timeoutSeconds(sdk, value);
  if (seconds === null) return false;
  return MILLISECOND_SDKS.has(String(sdk ?? '').trim().toLowerCase()) && seconds < 1;
}

/** The largest max_tokens that still finishes inside the ceiling. Pure. */
export function safeMaxTokens(tps = DEFAULT_TPS, wallClock = WALL_CLOCK, prefill = 0) {
  const rate = Number(tps || 0);
  const room = Math.max(0, Number(wallClock || 0) - Math.max(0, Number(prefill || 0)));
  if (rate <= 0) return 0;
  return Math.trunc(room * rate);
}

/**
 * Classify one call path against the clock. Pure. [state, detail].
 * The wall clock is checked before the client timeout, because a non-streaming
 * request past ten minutes fails on the far side whatever the client waits for.
 */
export function verdict(seconds, streams, timeoutS = null, wallClock = WALL_CLOCK, near = 0.8) {
  const shape = `${duration(seconds)} of generation estimated`;

  if (!streams && seconds > wallClock) {
    return ['over-wall-clock-not-streaming',
      `${shape} on a non-streaming path, past the ${duration(wallClock)} ` +
      'ceiling. That is a 504 timeout_error, or no response at all when an ' +
      'intermediate hop drops the idle connection first. Raising the client ' +
      'timeout does not move it.'];
  }
  if (timeoutS !== null && timeoutS !== undefined && seconds > timeoutS) {
    return ['over-client-timeout',
      `${shape} against a client timeout of ${duration(timeoutS)}, so the ` +
      'client gives up before the API is finished.'];
  }
  if (!streams && seconds >= wallClock * near) {
    return ['near-wall-clock-not-streaming',
      `${shape} on a non-streaming path, inside ${(near * 100).toFixed(0)}% of ` +
      `the ${duration(wallClock)} ceiling. One unusually long answer crosses it.`];
  }
  if (streams && seconds > wallClock) {
    return ['streams-past-ten-minutes',
      `${shape}, and the path streams, so the connection never goes idle and ` +
      'the ceiling does not apply. Worth the Message Batches API if nobody is ' +
      'waiting on it.'];
  }
  return ['within-budget', `${shape}.`];
}

function headers(key) {
  return { 'x-api-key': key, 'anthropic-version': VERSION,
           'content-type': 'application/json' };
}

async function get(key, path) {
  const res = await fetch(API + path, { headers: headers(key) });
  if (res.status === 401 || res.status === 403) {
    throw new Error(`${res.status} from Anthropic: ANTHROPIC_API_KEY has to be a workspace key`);
  }
  if (!res.ok) throw new Error(`${res.status} from ${path}`);
  return res.json();
}

/** The one non-GET call, and it neither creates nor bills anything. */
async function countInput(key, payloadPath) {
  const body = JSON.parse(await readFile(payloadPath, 'utf8'));
  const trimmed = Object.fromEntries(
    Object.entries(body).filter(([k]) => !SAMPLING_ONLY.has(k)));
  const res = await fetch(`${API}/messages/count_tokens`, {
    method: 'POST',  // count_tokens creates nothing and bills nothing
    headers: headers(key),
    body: JSON.stringify(trimmed),
  });
  if (!res.ok) throw new Error(`${res.status} from /messages/count_tokens`);
  return Math.trunc((await res.json())?.input_tokens ?? 0);
}

async function main() {
  const key = (process.env.ANTHROPIC_API_KEY || "dummy-anthropic-api-key");
  if (!key) {
    console.error('set ANTHROPIC_API_KEY to a workspace key');
    process.exitCode = 2;
    return;
  }
  const configPath = (process.env.CONFIG || "dummy-config") ?? process.argv[2];
  if (!configPath) {
    console.error('set CONFIG, or pass the call-paths JSON file as an argument');
    process.exitCode = 2;
    return;
  }
  const paths = JSON.parse(await readFile(configPath, 'utf8'));
  const tps = Number((process.env.TPS || "dummy-tps") ?? DEFAULT_TPS);
  const prefillTps = Number((process.env.PREFILL_TPS || "dummy-prefill-tps") ?? DEFAULT_PREFILL_TPS);
  const showAll = (process.env.SHOW_ALL || "dummy-show-all") === '1';

  const caps = new Map();
  let bad = 0;

  for (const name of Object.keys(paths).sort()) {
    const entry = paths[name] ?? {};
    const modelId = String(entry.model ?? '');
    const streams = Boolean(entry.stream);
    const sdk = entry.sdk;

    if (modelId && !caps.has(modelId)) {
      caps.set(modelId, (await get(key, `/models/${modelId}`)).max_tokens ?? null);
    }

    let inputTokens = Math.trunc(entry.input_tokens ?? 0);
    if (entry.payload) inputTokens = await countInput(key, entry.payload);

    const prefill = prefillSeconds(inputTokens, prefillTps);
    const seconds = prefill + generationSeconds(entry.max_tokens, tps);
    const client = timeoutSeconds(sdk, entry.timeout);

    const [state, detail] = verdict(seconds, streams, client);
    const line = `${state.padEnd(30)} ${name.padEnd(16)} ${detail}`;
    if (FINDINGS.has(state)) { bad += 1; console.warn(line); }
    else if (state === 'streams-past-ten-minutes') console.log(line);
    else if (showAll) console.log(line);

    if (unitSuspicion(sdk, entry.timeout)) {
      bad += 1;
      console.warn(`${'timeout-unit-mistake'.padEnd(30)} ${name.padEnd(16)} ` +
                   `timeout ${entry.timeout} on the ${sdk} client is ` +
                   `${(client ?? 0).toFixed(1)}s, not ${duration(entry.timeout ?? 0)}: ` +
                   'that unit is milliseconds');
    }

    if (state === 'over-wall-clock-not-streaming' || state === 'near-wall-clock-not-streaming') {
      console.warn(`  at ${tps.toFixed(0)} tok/s the largest max_tokens that ` +
                   `finishes inside the ceiling is ${safeMaxTokens(tps, WALL_CLOCK, prefill)}`);
      const cap = caps.get(modelId);
      if (cap) {
        console.warn(`  this model allows ${cap} output tokens, which is ` +
                     `${duration(generationSeconds(cap, tps))} on one call`);
      }
      console.warn('  repair: stream it. .stream() plus .finalMessage() returns the ' +
                   'identical Message object with no event handling, and the ' +
                   'connection never goes idle. For latency tolerant work use the ' +
                   'Message Batches API, which has no such clock. Printed, not applied.');
    }
  }

  console.log(`${Object.keys(paths).length} path(s) checked, ${bad} finding(s)`);
  process.exitCode = bad ? 1 : 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
